collections/map.go

```diff
@@ -0,0 +1,97 @@
+package collections
+
+import (
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+)
+
+// NewMap returns a Map given a StoreKey, a Prefix and the relative value and key encoders.
+func NewMap[K, V any](
+	sk storetypes.StoreKey, prefix Prefix,
+	keyEncoder KeyEncoder[K], valueEncoder ValueEncoder[V],
+) Map[K, V] {
+	return Map[K, V]{
+		kc:     keyEncoder,
+		vc:     valueEncoder,
+		sk:     sk,
+		prefix: prefix.Bytes(),
+	}
+}
+
+// Map represents the basic collections object.
+// It is used to map arbitrary keys to arbitrary
+// objects.
+type Map[K, V any] struct {
+	kc KeyEncoder[K]
+	vc ValueEncoder[V]
+
+	sk     storetypes.StoreKey
+	prefix []byte
+}
+
+// Set maps the provided value to the provided key in the store.
+func (m Map[K, V]) Set(ctx StorageProvider, key K, value V) {
+	keyBytes := m.encodeKey(key)
+
+	valueBytes, err := m.vc.Encode(value)
+	if err != nil {
+		panic(err)
+	}
+
+	m.getStore(ctx).Set(keyBytes, valueBytes)
+}
+
+// Get returns the value associated with the provided key,
+// or a ErrNotFound error in case the key is not present.
+func (m Map[K, V]) Get(ctx StorageProvider, key K) (V, error) {
+	keyBytes := m.encodeKey(key)
+
+	valueBytes := m.getStore(ctx).Get(keyBytes)
+	if valueBytes == nil {
+		var v V
+		return v, ErrNotFound{
+			HumanizedKey: m.kc.Stringify(key),
+			RawKey:       keyBytes,
+			ValueType:    m.vc.ValueType(),
+		}
+	}
+
+	v, err := m.vc.Decode(valueBytes)
+	if err != nil {
+		panic(err)
```

Comment:
No reason to panic here when you can return an error.

Reply:
This was something that I was really unsure about. Rationale for making it never failing is because in every module we end up using codec.MustMarshal/MustUnmarshal which yield panics in case of ser/de errors.

I am even wondering in which case a ser/de operation might fail, are they recoverable? 

For instance, if a deserialise operation fails, I would expect it to happen because there is incompatibility between bytes format and the type, which might caused by what? data corruption? pointing a collection towards a prefix which does not belong to it. these cases look un-recoverable to me, a sign of bad app/module setup.

I am open to changing the API to return and error, and open to extending this API with a MustGet/Set method and let everyone pick their poison.

Reply:
@aaronc do you have any feedback here? also @ValarDragon since you gave feedback on API ergonomics.

Reply:
panic says "somebody deferenced a nil pointer" or "divide by zero" or stuff like that, programmer errors. They're not meant to be used to express "these bytes didn't decode properly" or any other kind of application fault. The SDK gets this wrong basically everywhere so no shade to you. But yeah, if a function can fail, it should return an error.

Reply:
As I said I have no strong preference here, motivation is consistency, I was once told that it's better to be consistent even if it's not optimal than inconsistent.

Reply:
lets do error here. In the future we will think through bubbling up errors in a way that gives the app dev more control. We shouldn't decide to panic for them

Reply:
moved API to error, only 'bad ergonomics' is `Has` returning `bool, error`, I documented that err would be only on key encode failures.

Reply:
Ideally you want `Has(key) error` and to express "does not have" as a specific error return value.